brainframe_qt/api_utils/connection_manager.py

```diff
@@ -0,0 +1,154 @@
+        RUNTIME_ERROR = auto()
+
+    def __init__(self, *, parent: QObject):
+        super().__init__(parent=parent)
+
+        self._connection_state = self.ConnectionState.UNCONFIGURED
+        self._connection_configuration: Optional[ConnectionConfiguration] = None
+
+        self.server_settings = ServerSettings()
+
+        self._init_signals()
+
+    def run(self) -> None:
+        while not self.isInterruptionRequested():
+            if self.connection_state is self.ConnectionState.CONNECTED:
+                QThread.msleep(self._CONNECTED_SLEEP_TIME)
+            elif self.connection_state is self.ConnectionState.UNCONFIGURED:
+                self._get_configuration()
+            elif self.connection_state is self.ConnectionState.UNCONNECTED:
+                self._communicate_with_server()
+            elif self.connection_state in [
+                self.ConnectionState.LICENSE_UNVALIDATED,
+                self.ConnectionState.LICENSE_EXPIRED,
+                self.ConnectionState.LICENSE_INVALID,
+                self.ConnectionState.LICENSE_MISSING
+            ]:
+                self._validate_license()
+            elif self.connection_state is self.ConnectionState.RUNTIME_ERROR:
+                QThread.msleep(self._CONNECTED_SLEEP_TIME)
+            else:
+                exc = RuntimeError(f"Unknown ConnectionState {self.connection_state}")
+                self._handle_error(exc)
+
+    def _init_signals(self) -> None:
+        self.server_settings.value_changed.connect(self._handle_settings_change)
+
+    @property
+    def connection_state(self) -> ConnectionState:
+        return self._connection_state
+
+    @connection_state.setter
+    def connection_state(self, connection_state: ConnectionState) -> None:
+        self._connection_state = connection_state
+        self.connection_state_changed.emit(self._connection_state)
+
+    def invalidate_config(self) -> None:
+        """Force the ConnectionManager to restart the authentication process"""
+        self.connection_state = self.ConnectionState.UNCONFIGURED
+
+    def _get_configuration(self) -> None:
+        url = self.server_settings.server_url
+        username = self.server_settings.server_username
+        password = self.server_settings.server_password
+
+        if username and password:
+            password = decrypt(password)
+
+        self._connection_configuration = ConnectionConfiguration(
+            server_url=url,
+            server_username=username,
+            server_password=password,
+        )
+        api.set_url(self._connection_configuration.server_url)
+        api.set_credentials(self._connection_configuration.credentials)
+
+        self.connection_state = self.ConnectionState.UNCONNECTED
+
+    def _communicate_with_server(self) -> None:
+        try:
+            api.wait_for_server_initialization(self._CONNECT_TIMEOUT)
+        except (TimeoutError, bf_errors.ServerNotReadyError):
+            pass
```

Comment:
Should this set the state to UNCONNECTED?

Reply:
I think the only way to get to this code is if the state is already `UNCONNECTED`, so setting it to this state again wouldn't do anything. You could argue that it's more explicit though, and maybe more flexible to change? I don't have strong feelings about this.

Reply:
You know what, I think it might be good to set the state here. That would make this method more consistent with the others in the sense that the method always returns having asserted the state of the ConnectionManager.

Reply:
The disadvantage of always setting the value is that I'm constantly emitting the `connection_state_changed`. I could, however, de-duplicate the signal

Reply:
I think it would make sense to only emit that signal when the connection state has actually changed, not just when it's set.

Reply:
As it stands, with the current logic, it is only getting set when it's changing

Reply:
Right, but don't you think it's weird that setting the state triggers `connection_state_changed` when the connection state hasn't actually changed? That could be fixed by adding a check to that property.

Reply:
Then that would allow the state to be asserted here without spuriously triggering `connection_state_changed`.

Reply:
I made the change